Turn theme into a hugo module

deining opened the pull request:
This PR turns `hugo-fresh` into a hugo module. This way, the theme can pull in dependency `bulma` as go module.

With this PR in place, setting up a new site based on `hugo-fresh` can be done via:

```
hugo new site my-site
cd my-site
hugo mod init github.com/me/my-site
hugo mod get github.com/StefMa/hugo-fresh #optional
echo 'theme = "github.com/StefMa/hugo-fresh" >> config.toml
hugo serve
```

**Note:** This PR is work in progress. The concept works: with this PR applied, you can preview the exampleSite simply via `hugo server` (no `themesDir=../..` needed any more). If you like the idea, I'm willing to refine the implementation and add the documentation for the new route to install `hugo-fresh`.

StefMa commented:
Hey,

thanks for the contribution again!
I definitely like the idea.
Feel free to continue with the implementation.

On Sat, Mar 5, 2022, 8:41 AM Andreas Deininger ***@***.***>
wrote:

> This PR turns hugo-fresh into a hugo module. This way, the theme can pull
> in dependency bulma as go module.
>
> With this PR in place, setting up a new site based on hugo-fresh can be
> done via:
>
> hugo new site my-site
> cd my-site
> hugo mod init github.com/me/my-site
> hugo mod get github.com/StefMa/hugo-fresh #optional
> echo 'theme = "github.com/StefMa/hugo-fresh" >> config.toml
> hugo serve
>
> *Note:* This PR is work in progress. The concept works: with this PR
> applied, you can preview the exampleSite simply via hugo server (no
> themesDir=../.. needed any more). If you like the idea, I'm willing to
> refine the implementation and add the documentation for the new route to
> install hugo-fresh.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/StefMa/hugo-fresh/pull/134
> Commit Summary
>
>    - 2a0489b
>    <https://github.com/StefMa/hugo-fresh/pull/134/commits/2a0489bc8bcb11b48fbfd5faac3b93440335de50>
>    Turn theme into a hugo module
>
> File Changes
>
> (6 files <https://github.com/StefMa/hugo-fresh/pull/134/files>)
>
>    - *A* config.yaml
>    <https://github.com/StefMa/hugo-fresh/pull/134/files#diff-d8d0422389f03d783e32e627250fe29834bd09c6361640d1ff00661dd6820034>
>    (26)
>    - *M* exampleSite/config.yaml
>    <https://github.com/StefMa/hugo-fresh/pull/134/files#diff-1926bed36200257684beefb9e338d8dc0679b7226f1cb4290fb677139853dfb9>
>    (3)
>    - *A* exampleSite/go.mod
>    <https://github.com/StefMa/hugo-fresh/pull/134/files#diff-5bed2c958464666aa841aecfa747db471387119abc35bb3fbb729565aaac8818>
>    (7)
>    - *A* exampleSite/go.sum
>    <https://github.com/StefMa/hugo-fresh/pull/134/files#diff-f6be75f6781e07f5e910f0302efdca04702663d15147714c0d422cc7fadbf156>
>    (3)
>    - *A* go.mod
>    <https://github.com/StefMa/hugo-fresh/pull/134/files#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6>
>    (5)
>    - *A* go.sum
>    <https://github.com/StefMa/hugo-fresh/pull/134/files#diff-3295df7234525439d778f1b282d146a4f1ff6b415248aaac074e8042d9f42d63>
>    (2)
>
> Patch Links:
>
>    - https://github.com/StefMa/hugo-fresh/pull/134.patch
>    - https://github.com/StefMa/hugo-fresh/pull/134.diff
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/StefMa/hugo-fresh/pull/134>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOBQ67XP2IUIY24N5BTZR3U6MFY3ANCNFSM5P7K7MYA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

StefMa commented:
Hey @deining .
Are you still interested in this idea? 
What are the to-do`s to finish this up?
Camn I help you with someting?

deining commented:
> Hey @deining . Are you still interested in this idea?

Yes, I am. Sorry, this flew under my radar for a long time!

> What are the to-do`s to finish this up?

I realized that bulma is at version 0.9.4 meanwhile (my original PR was for 0.9.3). I corrected my PR, it now targets bulma 0.9.4.
So implementation wise, everything is ready to be rolled out :smile: I just removed the draft status from this PR and would like to see this merged soon.

> Camn I help you with someting?

Yes. CircleCi now fails with this message:

`Error: we found a go.mod file in your project, but you need to install Go to use it.`

Can you add `go` to the CI workflow, please?

Furthermore: With my PR in place, we have two installation approaches now: hugo module vs. git submodule. The problem with that is, that both approaches are out sync concerning `bulma` version: hugo module is at latest bulma version 0.9.4, while HEAD of repo is at 0.7.0. Both should be in sync however, as requested in #114 and addressed in #133 by me (unfortunately my PR is outdated now).

So we have two choices:
- update HEAD of repo to latest bulma version 0.9.4 (if wanted, can you take that on, please?)
- drop support for git submodule installations altogether and subsequently remove bulma files from repo

You are the repo owner, you have to decide which of the tweo proposed ways you want to go.

**Two more notes:**

1. Documentation wise, I find it strange that install instructions in `/docs/content/getstarted.md`, but neither in the exampleSite nor in the `https://hugo-fresh.vercel.app/` nor in `README.md`. This somehow hides away the whole documentation from the user. Was this done on purpose? If so, what was the motivation?
2. I realized that the repo contains parts of the resource cache in `resources` folder. According to this [comment](https://discourse.gohugo.io/t/why-does-hugo-create-the-resources-directory/13021/3), this is a good thing. If you update the bulma files inside this repo, the files inside this directory should be updated as well.

StefMa commented:
Thanks @deining and thank you for all the information you provide here 🙏 

Regarding CircleCI:
I would suggest we wait until we moved to GH Actions (https://github.com/StefMa/hugo-fresh/pull/157#issuecomment-1445823323). 

Regarding two versions (submodule and modules) I would go the "all-in hugo modules" way.
Let's drop submodules 🗑️.
Do you want to take care for this in this PR as well?

> Documentation wise, I find it strange that install instructions in /docs/content/getstarted.md, but neither in the exampleSite nor in the https://hugo-fresh.vercel.app/ nor in README.md. This somehow hides away the whole documentation from the user. Was this done on purpose? If so, what was the motivation?

Yep, I also realized this and probably move the documentation more prominent to the README.
Thanks for the hint 👍 I will think about it ...

> I realized that the repo contains parts of the resource cache in resources folder. According to this [comment](https://discourse.gohugo.io/t/why-does-hugo-create-the-resources-directory/13021/3), this is a good thing. If you update the bulma files inside this repo, the files inside this directory should be updated as well.

You mean I **should not be part of the repository**, right? 🤔 
If so, please don't hesitate to remove it in this PR as well 🙂

StefMa commented:
Thanks @deining and thank you for all the information you provide here 🙏 

Regarding CircleCI:
I would suggest we wait until we oved to GH Actions (https://github.com/StefMa/hugo-fresh/pull/157#issuecomment-1445823323). 

Regarding two versions (submodule and modules) I would go the "all-in hugo modules" way.
Let's drop submodules 🗑️.
Do you want to take care for this in this PR as well?

> Documentation wise, I find it strange that install instructions in /docs/content/getstarted.md, but neither in the exampleSite nor in the https://hugo-fresh.vercel.app/ nor in README.md. This somehow hides away the whole documentation from the user. Was this done on purpose? If so, what was the motivation?

Yep, I also realized this and probably move the documentation more prominent to the README.
Thanks for the hint 👍 I will think about it ...

> I realized that the repo contains parts of the resource cache in resources folder. According to this [comment](https://discourse.gohugo.io/t/why-does-hugo-create-the-resources-directory/13021/3), this is a good thing. If you update the bulma files inside this repo, the files inside this directory should be updated as well.

You mean it should **not be part of the repository**, right? 🤔 
If so, please don't hesitate to remove it in this PR as well 🙂

deining commented:
> Regarding two versions (submodule and modules) I would go the "all-in hugo modules" way. Let's drop submodules 🗑️.

Wise decision!

> Do you want to take care for this in this PR as well?

Yes, I will come back to this once migration to GitHub actions is done!

deining commented:
I removed all remnants of the git submodule + local bulma installation I could find.

> Yep, I also realized this and probably move the documentation more prominent to the README.

That's a goo idea and would, go ahead with this.

I'm done here too, handing over to your for approval/finalization.

deining committed:
Turn theme into a hugo module

This way, the theme pulls in dependency 'bulma' as go module

urban-adeininger committed:
Dropping bulma which is now pulled in as module

urban-adeininger committed:
Remove installation as git submodule (config, doc)

urban-adeininger committed:
Update generated resources (bulma 0.9.4)

StefMa committed:
Fix vercel deployment

StefMa committed:
Merge remote-tracking branch 'origin/master' into module

StefMa reviewed (approved):
Finally, we did it 🚀 
Thanks again @deining.
This moves the theme to the next level.

StefMa commented on config.yaml:
```diff
@@ -0,0 +1,28 @@
+# and mount points are declared
+module:
+  imports:
+  - path: github.com/jgthms/bulma
```
A follow-up question regarding this here @deining .

We now pull always the latest `main`/`master` branch from bulma right? 🤔 
Can we also specify a `git tag`? 🤔
Like  `github.com/jgthms/bulma@0.9.3`
I guess this is more stable than always pulling a develop branch?! 🤔

deining replied:
> We now pull always the latest `main`/`master` branch from bulma right?

No. we pull in `0.9.4` right now.

The bulma version used is defined in `go.mod`:

https://github.com/StefMa/hugo-fresh/blob/7987c63887d208977c332942cf7a9f756e96cc1d/go.mod#L5

To update bulma inside your theme use this command:

```bash
hugo mod get -u github.com/jgthms/bulma@9734c65c100   # commit hash
```

Instead of a commit, you can also use a version (the tag should have leading v though (e.g. v0.9.4)):

As an example, the command below can be used by users of your theme to update their theme version used:

```bash
hugo mod get -u github.com/StefMa/hugo-fresh@v0.2
```

This only works of course if you start using tags like vn.n. What about tagging this first module version with `v0.1` or so?

Hope that clarifies things!

StefMa replied:
Thank you very much.
Yes, that helped quite a lot.
Its sad that the `go.mod` didn't show the git tag but the hash instead 😭

deining replied:
> Its sad that the `go.mod` didn't show the git tag but the hash instead

If the name of the tag in the bulma repo were `v0.9.4` instead of `0.9.4`, you would see it.
So if you want that users of your theme see the tag instead of the hash, but better tag the inital module release with `v1.0`instead of `1.0`.

StefMa replied:
Thank you very much again 🙂 
Here we are: https://github.com/StefMa/hugo-fresh/releases/tag/v1.0.0

With that I have a go.mod and go.sum client like that:
```
// go.mod
require github.com/StefMa/hugo-fresh v1.0.0 // indirect

// go.sum
github.com/StefMa/hugo-fresh v1.0.0 h1:Nxo9WtzfBlaFi1G8NEHQLReJQz+Dn4jRxec+lMrPG6k=
github.com/StefMa/hugo-fresh v1.0.0/go.mod h1:yRzHyco0bfnITjvq7rfAiOADfGMrDlM0kXeKBllUJEc=
github.com/jgthms/bulma v0.0.0-20220508134905-3e00a8e6d0d0/go.mod h1:89FLBKXYFKfOKAoDeUT26V0pHGwzr205cMXAEqtAVOI=
```
